// pwmetrics/lib/metrics/metrics-adapter.js
"use strict";
// Copyright 2016 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("../utils/messages");
const logger_1 = require("../utils/logger");
const metrics_1 = require("./metrics");
const logger = logger_1.Logger.getInstance();
const checkMetrics = (metrics) => {
    const errorMessage = metrics.errorMessage;
    const explanation = metrics.details.explanation;
    if (errorMessage)
        logger.log(`${errorMessage} \n ${explanation}`);
};
const getMetricTitle = (metricId) => {
    try {
        return messages_1.getMessage(metricId);
    }
    catch (e) {
        return '';
    }
};
exports.adaptMetricsData = (res) => {
    const auditResults = res.audits;
    // has to be Record<string, LH.Audit.Result>
    const metricsAudit = auditResults.metrics;
    if (!metricsAudit || !metricsAudit.details || !metricsAudit.details.items)
        throw new Error('No metrics data');
    const metricsValues = metricsAudit.details.items[0];
    checkMetrics(metricsAudit);
    const colorP0 = 'yellow';
    const colorP2 = 'green';
    const colorVisual = 'blue';
    const timings = [];
    // @todo improve to Object.entries
    Object.keys(metricsValues).forEach(metricKey => {
        if (!Object.values(metrics_1.METRICS).includes(metricKey))
            return;
        const metricTitle = getMetricTitle(metricKey);
        const resolvedMetric = {
            title: metricTitle,
            id: metricKey,
            timing: metricsValues[metricKey],
            color: colorVisual
        };
        switch (metricKey) {
            case metrics_1.METRICS.TTFCP:
            case metrics_1.METRICS.TTFMP:
                resolvedMetric.color = colorP2;
                break;
            case metrics_1.METRICS.TTFCPUIDLE:
            case metrics_1.METRICS.TTI:
                resolvedMetric.color = colorP0;
                break;
        }
        timings.push(resolvedMetric);
    });
    return {
        timings,
        generatedTime: res.fetchTime,
        lighthouseVersion: res.lighthouseVersion,
        requestedUrl: res.requestedUrl,
        finalUrl: res.finalUrl,
    };
};
