// pwmetrics/lib/drive/gdrive.js
"use strict";
// Copyright 2016 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
Object.defineProperty(exports, "__esModule", { value: true });
const google = require('googleapis');
const promisify = require('micro-promisify');
const logger_1 = require("../utils/logger");
const GoogleOauth = require('../oauth/google-oauth');
const messages = require('../utils/messages');
const logger = logger_1.Logger.getInstance();
class GDrive {
    constructor(clientSecret) {
        this.clientSecret = clientSecret;
    }
    async getOauth() {
        try {
            if (this.oauth)
                return this.oauth;
            const googleOauth = new GoogleOauth();
            return this.oauth = await googleOauth.authenticate(this.clientSecret);
        }
        catch (error) {
            throw error;
        }
    }
    async uploadToDrive(data, fileName) {
        try {
            logger.log(messages.getMessage('G_DRIVE_UPLOADING'));
            const drive = google.drive({
                version: 'v3',
                auth: await this.getOauth()
            });
            const body = {
                resource: {
                    name: fileName,
                    mimeType: 'application/json',
                },
                media: {
                    mimeType: 'application/json',
                    body: JSON.stringify(data)
                }
            };
            const driveResponse = await promisify(drive.files.create)(body);
            await this.setSharingPermissions(driveResponse.id);
            logger.log(messages.getMessage('G_DRIVE_UPLOADED'));
            return driveResponse;
        }
        catch (error) {
            throw new Error(error);
        }
    }
    async setSharingPermissions(fileId) {
        try {
            const drive = google.drive({
                version: 'v3',
                auth: await this.getOauth()
            });
            const body = {
                resource: {
                    'type': 'anyone',
                    'role': 'writer'
                },
                fileId: fileId
            };
            return await promisify(drive.permissions.create)(body);
        }
        catch (error) {
            throw new Error(error);
        }
    }
}
module.exports = GDrive;
