// pwmetrics/lib/chart/chart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const wunderbar = require('@gribnoysup/wunderbar');
const eol = require('os').EOL;
const logger_1 = require("../utils/logger");
const logger = logger_1.Logger.getInstance();
exports.drawChart = (timings, options) => {
    const { lmargin, width, xlabel, xmin, xmax } = options;
    const normalizedTimings = timings.map(value => {
        return {
            value: value.timing,
            label: value.title,
            color: value.color
        };
    });
    const { __raw } = wunderbar(normalizedTimings, {
        min: xmin,
        max: xmax,
        length: width,
        format: '0,000'
    });
    const { normalizedValues, minValueFormatted, maxValueFormatted } = __raw;
    const yAxis = '│';
    const xAxis = '─';
    const corner = '╰';
    const padding = ' '.repeat(lmargin);
    const chart = normalizedValues
        .reverse()
        .map((value) => {
        const pad = lmargin - value.label.length;
        const paddedLabel = ' '.repeat(pad > 0 ? pad : 0) + value.label;
        return `${paddedLabel} ${yAxis}${value.coloredChartBar} ${value.formattedValue}`;
    })
        .join(`${eol}${padding} ${yAxis}${eol}`);
    const chartTop = `${padding} ${yAxis}`;
    const chartBottom = `${padding} ${corner}${xAxis.repeat(width)}`;
    const labelPadding = ' '.repeat(Math.max(0, (width - xlabel.length - 2) / 2));
    const chartScale = `${padding}  ${minValueFormatted}${labelPadding}${xlabel}${labelPadding}${maxValueFormatted}`;
    logger.log('');
    logger.log([chartTop, chart, chartBottom, chartScale].join(eol));
    logger.log('');
};
