// pwmetrics/lib/index.js
"use strict";
// Copyright 2016 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
const opn = require('open');
const os = require('os');
const path = require('path');
const metrics_1 = require("./metrics/metrics");
const logger_1 = require("./utils/logger");
const lh_runner_1 = require("./lh-runner");
const sheets_1 = require("./sheets");
const metrics_adapter_1 = require("./metrics/metrics-adapter");
const expectations_1 = require("./expectations");
const upload_1 = require("./upload");
const fs_1 = require("./utils/fs");
const messages_1 = require("./utils/messages");
const chart_1 = require("./chart/chart");
const getTimelineViewerUrl = (id) => `https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://drive.google.com/file/d//${id}/view?usp=drivesdk`;
class PWMetrics {
    constructor(url, opts) {
        this.url = url;
        this.flags = {
            runs: 1,
            submit: false,
            upload: false,
            view: false,
            expectations: false,
            json: false,
            chromeFlags: '',
            showOutput: true,
            failOnError: false,
            outputPath: 'stdout',
        };
        this.flags = Object.assign({}, this.flags, opts.flags);
        this.runs = this.flags.runs;
        this.sheets = opts.sheets;
        this.clientSecret = opts.clientSecret;
        const expectations = opts.expectations;
        // normalize path if provided
        if (this.flags.chromePath) {
            this.flags.chromePath = path.normalize(this.flags.chromePath);
        }
        if (this.flags.expectations) {
            if (expectations) {
                expectations_1.validateMetrics(expectations);
                this.normalizedExpectations = expectations_1.normalizeExpectationMetrics(expectations);
            }
            else
                throw new Error(messages_1.getMessageWithPrefix('ERROR', 'NO_EXPECTATIONS_FOUND'));
        }
        this.logger = logger_1.Logger.getInstance({ showOutput: this.flags.showOutput });
    }
    async start() {
        const runs = Array.apply(null, { length: +this.runs }).map(Number.call, Number);
        let metricsResults = [];
        for (let runIndex of runs) {
            try {
                const lhRunner = new lh_runner_1.LHRunner(this.url, this.flags);
                const lhTrace = await lhRunner.run();
                metricsResults[runIndex] = await this.recordLighthouseTrace(lhTrace);
                this.logger.log(messages_1.getMessageWithPrefix('SUCCESS', 'SUCCESS_RUN', runIndex, runs.length));
            }
            catch (error) {
                metricsResults[runIndex] = error;
                this.logger.error(messages_1.getMessageWithPrefix('ERROR', 'FAILED_RUN', runIndex, runs.length, error.message));
            }
        }
        const results = { runs: metricsResults.filter(r => !(r instanceof Error)) };
        if (this.runs > 1 && !this.flags.submit) {
            results.median = this.findMedianRun(results.runs);
            this.logger.log(messages_1.getMessage('MEDIAN_RUN'));
            this.displayOutput(results.median);
        }
        else if (this.flags.submit) {
            const sheets = new sheets_1.Sheets(this.sheets, this.clientSecret);
            if (this.sheets.options.uploadMedian) {
                results.median = this.findMedianRun(results.runs);
                await sheets.appendResults([results.median]);
            }
            else {
                await sheets.appendResults(results.runs);
            }
        }
        await this.outputData(results);
        if (this.flags.expectations) {
            const resultsToCompare = this.runs > 1 ? results.median.timings : results.runs[0].timings;
            const hasExpectationsWarnings = this.resultHasExpectationIssues(resultsToCompare, 'warn');
            const hasExpectationsErrors = this.resultHasExpectationIssues(resultsToCompare, 'error');
            if (hasExpectationsWarnings || hasExpectationsErrors) {
                expectations_1.checkExpectations(resultsToCompare, this.normalizedExpectations);
                if (hasExpectationsErrors && this.flags.failOnError) {
                    throw new Error(messages_1.getMessage('HAS_EXPECTATION_ERRORS'));
                }
                else {
                    this.logger.warn(messages_1.getMessage('HAS_EXPECTATION_ERRORS'));
                }
            }
        }
        return results;
    }
    resultHasExpectationIssues(timings, issueType) {
        return timings.some((timing) => {
            const expectation = this.normalizedExpectations[timing.id];
            if (!expectation) {
                return false;
            }
            const expectedLimit = expectation[issueType];
            return expectedLimit !== undefined && timing.timing >= expectedLimit;
        });
    }
    async recordLighthouseTrace(data) {
        try {
            const preparedData = metrics_adapter_1.adaptMetricsData(data.lhr);
            if (this.flags.upload) {
                const driveResponse = await upload_1.upload(data, this.clientSecret);
                this.view(driveResponse.id);
            }
            if (!this.flags.submit && this.runs <= 1) {
                this.displayOutput(preparedData);
            }
            return preparedData;
        }
        catch (error) {
            throw error;
        }
    }
    displayOutput(data) {
        if (!this.flags.json)
            this.showChart(data);
        return data;
    }
    showChart(data) {
        // reverse to preserve the order, because cli-chart.
        let timings = data.timings;
        timings = timings.filter(r => {
            // filter out metrics that failed to record
            if (r.timing === undefined || isNaN(r.timing)) {
                this.logger.error(messages_1.getMessageWithPrefix('ERROR', 'METRIC_IS_UNAVAILABLE', r.title));
                return false;
            }
            else {
                return true;
            }
        });
        const fullWidthInMs = Math.max(...timings.map(result => result.timing));
        const maxLabelWidth = Math.max(...timings.map(result => result.title.length));
        const terminalWidth = +process.stdout.columns || 90;
        chart_1.drawChart(timings, {
            // 90% of terminal width to give some right margin
            width: terminalWidth * 0.9 - maxLabelWidth,
            xlabel: 'Time (ms) since navigation start',
            xmin: 0,
            // nearest second
            xmax: Math.ceil(fullWidthInMs / 1000) * 1000,
            lmargin: maxLabelWidth + 1,
        });
        return data;
    }
    findMedianRun(results) {
        const TTFCPUIDLEValues = results.map(r => r.timings.find(timing => timing.id === metrics_1.METRICS.TTFCPUIDLE).timing);
        const medianTTFCPUIDLE = this.median(TTFCPUIDLEValues);
        // in the case of duplicate runs having the exact same TTFI, we naively pick the first
        // @fixme, but any for now...
        return results.find((result) => result.timings.find((timing) => timing.id === metrics_1.METRICS.TTFCPUIDLE && timing.timing === medianTTFCPUIDLE));
    }
    median(values) {
        if (values.length === 1)
            return values[0];
        values.sort((a, b) => a - b);
        const half = Math.floor(values.length / 2);
        return values[half];
    }
    view(id) {
        if (this.flags.view) {
            opn(getTimelineViewerUrl(id));
        }
    }
    outputData(data) {
        if (this.flags.json) {
            // serialize accordingly
            const formattedData = JSON.stringify(data, null, 2) + os.EOL;
            // output to file.
            if (this.flags.outputPath !== 'stdout') {
                return fs_1.writeToDisk(this.flags.outputPath, formattedData);
                // output to stdout
            }
            else if (formattedData) {
                return Promise.resolve(process.stdout.write(formattedData));
            }
        }
    }
}
module.exports = PWMetrics;
