// pwmetrics/lib/lh-runner.js
"use strict";
// Copyright 2018 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
Object.defineProperty(exports, "__esModule", { value: true });
const lighthouse = require('lighthouse/lighthouse-core');
const parseChromeFlags = require('lighthouse/lighthouse-cli/run').parseChromeFlags;
const chrome_launcher_1 = require("chrome-launcher");
const logger_1 = require("./utils/logger");
const messages_1 = require("./utils/messages");
const perfConfig = require('./perf-config');
const MAX_LIGHTHOUSE_TRIES = 2;
class LHRunner {
    //@todo improve FeatureFlags -> LHFlags
    constructor(url, flags) {
        this.url = url;
        this.flags = flags;
        this.tryLighthouseCounter = 0;
        this.logger = logger_1.Logger.getInstance({ showOutput: this.flags.showOutput });
    }
    async run() {
        try {
            let lhResults;
            await this.launchChrome();
            if (process.env.CI) {
                // handling CRI_TIMEOUT issue - https://github.com/GoogleChrome/lighthouse/issues/833
                this.tryLighthouseCounter = 0;
                lhResults = await this.runLighthouseOnCI().then((lhResults) => {
                    // fix for https://github.com/paulirish/pwmetrics/issues/63
                    return new Promise(resolve => {
                        this.logger.log(messages_1.getMessage('WAITING'));
                        setTimeout(_ => {
                            return resolve(lhResults);
                        }, 2000);
                    });
                });
            }
            else {
                lhResults = await lighthouse(this.url, this.flags, perfConfig);
            }
            await this.killLauncher();
            return lhResults;
        }
        catch (error) {
            await this.killLauncher();
            throw error;
        }
    }
    async killLauncher() {
        if (typeof this.launcher !== 'undefined') {
            await this.launcher.kill();
        }
    }
    async runLighthouseOnCI() {
        try {
            return await lighthouse(this.url, this.flags, perfConfig);
        }
        catch (error) {
            if (error.code === 'CRI_TIMEOUT' && this.tryLighthouseCounter <= MAX_LIGHTHOUSE_TRIES) {
                return await this.retryLighthouseOnCI();
            }
            if (this.tryLighthouseCounter > MAX_LIGHTHOUSE_TRIES) {
                throw new Error(messages_1.getMessage('CRI_TIMEOUT_REJECT'));
            }
        }
    }
    async retryLighthouseOnCI() {
        this.tryLighthouseCounter++;
        this.logger.log(messages_1.getMessage('CRI_TIMEOUT_RELAUNCH'));
        try {
            return await this.runLighthouseOnCI();
        }
        catch (error) {
            this.logger.error(error.message);
            this.logger.error(messages_1.getMessage('CLOSING_CHROME'));
            await this.killLauncher();
        }
    }
    async launchChrome() {
        try {
            this.logger.log(messages_1.getMessage('LAUNCHING_CHROME'));
            this.launcher = await chrome_launcher_1.launch({
                port: this.flags.port,
                chromeFlags: parseChromeFlags(this.flags.chromeFlags),
                chromePath: this.flags.chromePath
            });
            this.flags.port = this.launcher.port;
            return this.launcher;
        }
        catch (error) {
            this.logger.error(error);
            await this.killLauncher();
            return error;
        }
    }
}
exports.LHRunner = LHRunner;
