// pwmetrics/lib/sheets/gsheets.js
"use strict";
// Copyright 2016 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
Object.defineProperty(exports, "__esModule", { value: true });
const google = require('googleapis');
const promisify = require('micro-promisify');
const { getMessage } = require('../utils/messages');
const logger_1 = require("../utils/logger");
const logger = logger_1.Logger.getInstance();
async function getRange(auth, range, spreadsheetId) {
    try {
        const sheets = google.sheets('v4');
        const response = await promisify(sheets.spreadsheets.values.get)({
            auth: auth,
            spreadsheetId: spreadsheetId,
            range: range
        });
        return response.values;
    }
    catch (error) {
        logger.error(getMessage('G_SHEETS_API_ERROR', error));
        throw new Error(error);
    }
}
const formatValues = (values) => {
    let newValues = values.slice();
    return newValues.reduce((result, value) => {
        return result.concat(value.slice(3).join('\t')).concat('\n');
    }, []).join('');
};
async function appendResults(auth, valuesToAppend, options) {
    try {
        const sheets = google.sheets('v4');
        // clone values to append
        const values = Object.assign([], valuesToAppend);
        logger.log(getMessage('G_SHEETS_APPENDING', formatValues(valuesToAppend)));
        const response = await promisify(sheets.spreadsheets.values.append)({
            auth: auth,
            spreadsheetId: options.spreadsheetId,
            range: `${options.tableName}!A1:C1`,
            valueInputOption: 'USER_ENTERED',
            resource: {
                values: values,
            },
        });
        const rangeValues = await getRange(auth, response.updates.updatedRange, options.spreadsheetId);
        logger.log(getMessage('G_SHEETS_APPENDED', formatValues(rangeValues)));
    }
    catch (error) {
        logger.error(getMessage('G_SHEETS_API_ERROR', error));
        throw new Error(error);
    }
}
exports.appendResults = appendResults;
