// pwmetrics/lib/sheets/index.js
"use strict";
// Copyright 2016 Google Inc. All Rights Reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE
Object.defineProperty(exports, "__esModule", { value: true });
const GoogleOauth = require('../oauth/google-oauth');
const gsheets = require("./gsheets");
// @todo add 'import' after moving all stuff to typescript
const { getMessage } = require('../utils/messages');
const { METRICS } = require('../metrics/metrics');
const SHEET_TYPES = {
    'GOOGLE_SHEETS': 'GOOGLE_SHEETS'
};
class Sheets {
    constructor(config, clientSecret) {
        this.config = config;
        this.clientSecret = clientSecret;
        this.validateOptions(config, clientSecret);
    }
    validateOptions(config, clientSecret) {
        if (!config || !Object.keys(config).length)
            throw new Error(getMessage('NO_GOOGLE_SHEET_OPTIONS'));
        const sheetType = config.type;
        if (!Object.keys(SHEET_TYPES).includes(sheetType)) {
            throw new Error(getMessage('NO_SHEET_TYPE', sheetType));
        }
        switch (sheetType) {
            case SHEET_TYPES.GOOGLE_SHEETS:
                if (!config.options.spreadsheetId || !config.options.tableName || !clientSecret)
                    throw new Error(getMessage('NO_GOOGLE_SHEET_OPTIONS'));
                break;
            default:
                throw new Error(getMessage('NO_GOOGLE_SHEET_OPTIONS'));
        }
    }
    appendResults(results) {
        switch (this.config.type) {
            case SHEET_TYPES.GOOGLE_SHEETS:
                return this.appendResultsToGSheets(results);
        }
    }
    async appendResultsToGSheets(results) {
        let valuesToAppend = [];
        results.forEach(data => {
            const getTiming = (key) => data.timings.find(t => t.id === key).timing;
            const dateObj = new Date(data.generatedTime);
            // order matters
            valuesToAppend.push([
                data.lighthouseVersion,
                data.requestedUrl,
                `${dateObj.toLocaleDateString()} ${dateObj.toLocaleTimeString()}`,
                getTiming(METRICS.TTFCP),
                getTiming(METRICS.TTFMP),
                getTiming(METRICS.SI),
                getTiming(METRICS.TTFCPUIDLE),
                getTiming(METRICS.TTI),
            ]);
        });
        try {
            const googleOauth = new GoogleOauth();
            const oauth = await googleOauth.authenticate(this.clientSecret);
            await gsheets.appendResults(oauth, valuesToAppend, this.config.options);
        }
        catch (error) {
            throw error;
        }
    }
}
exports.Sheets = Sheets;
